controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                return false;
+            }
+
+            reporting_agent_al_mac = TxLinkMetricData->reporter_al_mac();
+            neighbor_al_mac        = TxLinkMetricData->neighbor_al_mac();
+
+            LOG(DEBUG) << "recieved  tlvTransmitterLinkMetric from al_mac ="
+                       << network_utils::mac_to_string(reporting_agent_al_mac) << std::endl
+                       << "reported  al_mac =" << network_utils::mac_to_string(neighbor_al_mac)
+                       << std::endl;
+
+            //fill tx data from TLV
+            if (!new_link_metric_data.add_transmitter_link_metric(TxLinkMetricData)) {
+                LOG(ERROR) << "adding TxLinkMetricData has failed";
+                return false;
+            }
+        } else if (tlvType == int(ieee1905_1::eTlvType::TLV_RECEIVER_LINK_METRIC)) {
+            // parse tlvReceiverLinkMetric
+            auto RxLinkMetricData = cmdu_rx.addClass<ieee1905_1::tlvReceiverLinkMetric>();
+            if (!RxLinkMetricData) {
+                LOG(ERROR) << "addClass ieee1905_1::tlvReceiverLinkMetric has failed";
+                return false;
+            }
+
+            if (reporting_agent_al_mac != RxLinkMetricData->reporter_al_mac()) {
+                LOG(ERROR) << "TLV_RECEIVER_LINK_METRIC reporter al_mac ="
+                           << network_utils::mac_to_string(reporting_agent_al_mac) << std::endl
+                           << " and TLV_TRANSMITTER_LINK_METRIC reporter al_mac ="
+                           << network_utils::mac_to_string(RxLinkMetricData->reporter_al_mac())
+                           << std::endl
+                           << " not the same";
+                return false;
+            }
+
+            if (neighbor_al_mac != RxLinkMetricData->neighbor_al_mac()) {
+                LOG(ERROR) << "TLV_RECEIVER_LINK_METRIC reported al_mac ="
+                           << network_utils::mac_to_string(neighbor_al_mac) << std::endl
+                           << " and TLV_TRANSMITTER_LINK_METRIC reported al_mac ="
+                           << network_utils::mac_to_string(RxLinkMetricData->neighbor_al_mac())
+                           << std::endl
+                           << " not the same";
+                return false;
+            }
+
+            LOG(DEBUG) << "recieved  tlvReceiverLinkMetric from al_mac="
+                       << network_utils::mac_to_string(reporting_agent_al_mac) << std::endl
+                       << "reported  al_mac =" << network_utils::mac_to_string(neighbor_al_mac)
+                       << std::endl;
+
+            //fill rx data from TLV
+            if (!new_link_metric_data.add_receiver_link_metric(RxLinkMetricData)) {
+                LOG(ERROR) << "adding RxLinkMetricData has failed";
+                return false;
+            }
+        } else {
+            LOG(ERROR) << " unexpected tlv is recieved ";
+            return false;
+        }
+    }
+
```

Comment:
not sure why do we need the assumption , the  map<map<link_metrics>> solution supports all combination. no sure what is so special in a case where agent asked not to report all his neighbors (we would add it to the report anyways).

For the second part , you are talking about the actual implementation.
I'm don't think we need to cover disappearing agent scenarios for certification.
but adding clear() method is not a bad idea .
So basically you prefer clearing all reporting agent's map every time we get new link metric from him.

p.s
i will  erase all link/ap metric db  after combined_metrics message is sent . please correct me if i'm wrong.

Reply:
> not sure why do we need the assumption , the map<map<link_metrics>> solution supports all combination. no sure what is so special in a case where agent asked not to report all his neighbors (we would add it to the report anyways).

Sorry, that statement referred to the removal of stale neighbours. To be able to remove stale neighbours, we have to know *when* a neighbour is stale. Since we always ask for all metrics, not for a specific neighbour, we know that when we get the metrics, it's for all neighbours. Therefore, we can simply clear all the existing metrics for that particular reporting agent.


> So basically you prefer clearing all reporting agent's map every time we get new link metric from him.

Yep.


> i will erase all link/ap metric db after combined_metrics message is sent . please correct me if i'm wrong.

That is wrong. In general, it is possible that you receive new metrics for one agent but not another, so you should only clear the old metrics when new metrics are actually received.

Reply:
ok.
I'll add the clear()  method and use it each time new metrics arrives (need to run the test again to see that it holds though).
will add some comments to explain this in the code.